src/bulls_and_cows_lib/board.cpp

```diff
@@ -1,8 +1,185 @@
 
+
 #include "board.hpp"
+#include "random.hpp"
+#include <string>
+using namespace std; // lifehack
 
 namespace bulls_and_cows {
 
-    // TODO: define the body of the functions declared in board.cpp
+    Board create_board(const GameOptions& game_options)
+    {
+        Board b; // creer une structure board vide
+        for (size_t i = 0; i < game_options.number_of_characters_per_code;
+             i++) // boucle for qui remplit le code secret d'une taille game_options.number_of_characters_per_code
+        {
+            b.secret_code.value += generate_random_character(
+                game_options.minimum_allowed_character,
+                game_options
+                    .maximum_allowed_character); // lettre entre la lettre min et max de l'option passé en parametre
+        }
+
+        // TODO: peut etre initialiser l array de feedback
+        return b;
+    }
+
+    bool validate_attempt(const GameOptions& game_options, const Code& attempt)
+    {
+        string codex = attempt.value;
+        if (codex.length() != game_options.number_of_characters_per_code) // verification de la taille du code
+        {
+            return false;
+        }
+
+        for (int i = 0; i < codex.size(); i++) //  on parcourt de chaque caracter du codex
+        {
+            if (codex[i] < game_options.minimum_allowed_character ||
+                codex[i] > game_options.maximum_allowed_character) // si une des lettres est en dehors des bornes fixés
+                                                                   // (genre 'A' et 'H')
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    Feedback compare_attempt_with_secret_code(const Code& attempt, const Code& secret_code)
+    {
+        string REALCODE = secret_code.value;
+        string codex = attempt.value;
+        Feedback f;
+        int cows = 0;
+        int bulls = 0;
+
+        //on crééer un tableau de boolean
+        vector<bool> bullvector;
+        for (int i = 0; i < codex.size(); i++)
+        {
+            bullvector.push_back(false);
+        }
+
+        // on compte les bulls
+        for (int i = 0; i < codex.size(); i++) // on parcourt de chaque caracter du codex
+        {
+            if (codex[i] == REALCODE[i]) // si on a trouvé le caractere dans le code
+            {
+                bulls++;
+                bullvector[i] = true;
+            }
+        }
+
+        // on compte les cows
+        // !!!!!! on part du principe que si dans l'attempt on met 2 fois un char présent 1 fois dans le secret_code,
+        // alors on a 2 cows. <- FAUX
+        for (int i = 0; i < codex.size(); i++) // on parcourt de chaque caracter du codex
+        {
```

Comment:
tu appelles potentiellement 2 fois REALCODE.find(codex[i]), tu devrais stocker son résultat dans une variable found_realcode_char_index, au premier appel.
Et je pense que ta condition n'est pas exacte, il faut je pense comparer found_realcode_char_index avec i (que tu devrais renommer attempt_char_index pour + de clareté)...
mais t'es pas loin

Reply:
C'est noté je modifie ça.